backend/src/api/datasource/datasource.controller.ts

```diff
@@ -0,0 +1,187 @@
+import {
+  Controller,
+  Delete,
+  Get,
+  Inject,
+  Param,
+  Req,
+  Res,
+  UseGuards,
+} from '@nestjs/common';
+import { FitnessRepository } from '../../db/repositories/fitness.repository';
+import { NestRequest } from '../../types/request.type';
+import { AutoGuard } from '../../auth/auto.guard';
+import { Response } from 'express';
+import { FitnessService } from '../../integration/fitness/fitness.service';
+import { FitBitProvider } from '../../integration/fitness/providers/fitbit.provider';
+import { LOGGER_SERVICE, LoggerService } from '../../logger/logger.service';
+
+@Controller('datasource')
+export class DatasourceController {
+  constructor(
+    private fitnessRepository: FitnessRepository,
+    private fitnessService: FitnessService,
+    @Inject(LOGGER_SERVICE)
+    private loggerService: LoggerService,
+  ) {}
+
+  @Get()
+  @UseGuards(AutoGuard)
+  public async getDatasourcesForUser(@Req() request: NestRequest) {
+    const datasourcesForUser = await this.fitnessService.getDatasourcesForUser(
+      request.user.id,
+    );
+
+    return datasourcesForUser.map((provider) => provider.getInfo());
+  }
+
+  @Get('/:id')
+  @UseGuards(AutoGuard)
+  public async getDatasourceForUser(
+    @Req() request: NestRequest,
+    @Param() params: any,
+  ) {
+    const datasource = await this.fitnessService.getProviderForUserById(
+      request.user.id,
+      params.id,
+    );
+
+    return datasource?.getInfo();
+  }
+
+  @Delete('/:id')
+  @UseGuards(AutoGuard)
+  public async deleteDatasource(
+    @Req() request: NestRequest,
+    @Param() params: any,
+    @Res() res: Response,
+  ) {
+    // Verify if the datasource exists
+    const datasource = await this.fitnessRepository.getProviderForUserById(
+      request.user.id,
+      params.id,
+    );
+
+    if (!datasource) {
+      res.status(400).json({
+        error: 'Datasource was not enabled',
+      });
+      return;
+    }
+
+    await this.fitnessRepository.deleteProvider({
+      where: { userId: request.user.id, type: params.id },
+    });
+
+    res.status(204);
+  }
+
+  @Get('/:id/authorize')
+  @UseGuards(AutoGuard)
+  public async getAuthorizeURL(
+    @Req() request: NestRequest,
+    @Param() params: any,
+    @Res() response: Response,
+  ) {
+    const responsibleProvider =
+      await this.fitnessService.getProviderForUserById(
+        request.user.id,
+        params.id,
+      );
+
+    if (!responsibleProvider) {
+      response.status(400).json({
+        error: 'Provider not available',
+      });
+      return;
```

Comment:
Again, in the name of deduplication, this snippet could be made reusable.

Reply:
This is a single line, do you really think this make sense for the 4 occurrences?

Reply:
Performance wise, it makes no difference. I think readability would benefit from it, but we can discuss this offline.